pypit/spectrographs/bpmimage.py

```diff
@@ -0,0 +1,175 @@
+# Module for generating the BPM image
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import os
+
+from pypit import msgs
+from pypit.core import arprocimg
+from pypit.spectrographs import lris
+from pypit.spectrographs import deimos
+
+from pypit import ardebug as debugger
+
+# For out of PYPIT running
+if msgs._debug is None:
+    debug = debugger.init()
+    debug['develop'] = True
+    msgs.reset(debug=debug, verbosity=2)
+
+# Does not need to be global, but I prefer it
+frametype = 'bpm'
+
+
+class BPMImage(object):
+    """
+    This class is primarily designed to generate an Bad Pixel Image
+      The master() method will return the image
+
+    There are several ways to build a BPM:
+       1. keck_lris_red
+          Provide binning and detector number to generate this instrument specific BPM
+       2. keck_deimos
+          Provide detector number to generate this instrument specific BPM
+       3. From a bias image (not well tested)
+          Set reduce_badpix='bias'
+          Provide msbias image
+       4. Dummy image from shape
+
+    Parameters
+    ----------
+    spectrograph : str (optional)
+       Used to specify properties of the detector (for processing)
+       Attempt to set with settings['run']['spectrograph'] if not input
+    settings : dict (optional)
+      Settings for trace slits
+    det : int, optional
+      Detector index
+      Required for LRISr and DEIMOS
+    binning : Table (optional)
+        Could remove if we put binning in settings['detector']
+    shape : tuple
+      Image shape;  used to construct a dummy BPM if all else fails
+    msbias : ndarray, optional
+      Used to construct the BPM if reduce_badpix=='bias'
+    reduce_badpix : str, optional
+      'bias' -- Build from bias images
+
+    Attributes
+    ----------
+    bpm : ndarray
+
+    """
+    # Keep order same as processimages (or else!)
+    def __init__(self, spectrograph=None, settings=None, det=None,
+                 binning=None, reduce_badpix=None, msbias=None, shape=None):
+
+        # Parameters unique to this Object
+        self.spectrograph = spectrograph
+        self.binning = binning
+        self.det = det
+
+        self.reduce_badpix = reduce_badpix
+        self.msbias = msbias
+
+        self.shape = shape
+        self.settings = settings
+
+        # Checks
+        if (self.reduce_badpix == 'bias') and (self.msbias is None):
+            msgs.error("Need to supply msbias image with this option")
+        if (self.spectrograph == 'keck_deimos') and (self.det is None):
+            msgs.error("Need to supply det with this option")
+        if (self.spectrograph == 'keck_lris_red') and (self.det is None):
```

Comment:
Maybe put a TODO in here, so we can come back to this later once the Spectrograph class ideas are sorted out.

Reply:
added